pkg/services/install/install_test.go

```diff
@@ -0,0 +1,524 @@
+	var fakeGitProvider *gitprovidersfakes.FakeGitProvider
+	var repoWriter gitopswriter.RepoWriter
+	var log logger.Logger
+	var testNamespace string
+	var configRepo gitproviders.RepoURL
+	var err error
+	const clusterName = "test-cluster"
+	var _ = BeforeEach(func() {
+		testNamespace = "test-namespace"
+		configRepo, err = gitproviders.NewRepoURL("ssh://git@github.com/test-user/test-repo")
+		Expect(err).ShouldNot(HaveOccurred())
+		fakeFluxClient = &fluxfakes.FakeFlux{}
+		fakeKubeClient = &kubefakes.FakeKube{}
+		fakeGitClient = &gitfakes.FakeGit{}
+		fakeGitProvider = &gitprovidersfakes.FakeGitProvider{}
+		log = &loggerfakes.FakeLogger{}
+		repoWriter = gitopswriter.NewRepoWriter(log, fakeGitClient, fakeGitProvider)
+		installer = NewInstaller(fakeFluxClient, fakeKubeClient, fakeGitClient, fakeGitProvider, log, repoWriter)
+	})
+
+	Context("error paths", func() {
+		someError := errors.New("some error")
+
+		It("should fail validating wego installation", func() {
+			fakeKubeClient.GetClusterStatusReturns(kube.Unknown)
+
+			err := installer.Install(testNamespace, configRepo, true)
+			Expect(err).Should(MatchError("failed validating wego installation: Weave GitOps cannot talk to the cluster"))
+		})
+
+		It("should fail getting cluster name", func() {
+			fakeKubeClient.GetClusterStatusReturns(kube.Unmodified)
+			fakeKubeClient.GetWegoConfigReturns(&kube.WegoConfig{
+				FluxNamespace: testNamespace,
+				WegoNamespace: testNamespace,
+			}, nil)
+			fakeKubeClient.GetClusterNameReturns("", someError)
+
+			err := installer.Install(testNamespace, configRepo, true)
+			Expect(err).Should(MatchError(fmt.Sprintf("failed getting cluster name: %s", someError)))
+		})
+
+		It("should fail installing flux", func() {
+			fakeKubeClient.GetClusterStatusReturns(kube.Unmodified)
+			fakeKubeClient.GetWegoConfigReturns(&kube.WegoConfig{
+				FluxNamespace: testNamespace,
+				WegoNamespace: testNamespace,
+			}, nil)
+			fakeKubeClient.GetClusterNameReturns(clusterName, nil)
+
+			fakeFluxClient.InstallReturns(nil, someError)
+
+			err := installer.Install(testNamespace, configRepo, true)
+			Expect(err).Should(MatchError(fmt.Sprintf("failed installing flux: %s", someError)))
+		})
+
+		It("should fail getting bootstrap manifests", func() {
+			fakeKubeClient.GetClusterStatusReturns(kube.Unmodified)
+			fakeKubeClient.GetWegoConfigReturns(&kube.WegoConfig{
+				FluxNamespace: testNamespace,
+				WegoNamespace: testNamespace,
+			}, nil)
+			fakeKubeClient.GetClusterNameReturns(clusterName, nil)
+
+			fakeFluxClient.InstallReturnsOnCall(0, nil, nil)
+			fakeFluxClient.InstallReturnsOnCall(1, nil, someError)
+
+			err := installer.Install(testNamespace, configRepo, true)
+			Expect(err).Should(MatchError(fmt.Sprintf("failed getting bootstrap manifests: failed getting runtime manifests: %s", someError)))
+		})
+
+		It("should fail getting default branch", func() {
+			fakeKubeClient.GetClusterStatusReturns(kube.Unmodified)
```

Comment:
These 8 lines appear to be repeated in quite of few of these tests, consider either move to setup or make a function that each could call.

Reply:
@palemtnrider  Can't do that. Each test uses specific setup lines. Not all lines appear in all tests and not all are the same. Let me know if I'm missing something. This another approach where I would be duplicate any single line but seems a bit confusing. The huge advantage I see is that if I need to make a change somewhere it would be in a single place only. Let me know if you want to discuss this in a call.

```
Context("error paths2", func() {
		someError := errors.New("some error")
		It("should fail validating wego installation", func() {
			fakeKubeClient.GetClusterStatusReturns(kube.Unknown)

			err := installer.Install(testNamespace, configRepo, true)
			Expect(err).Should(MatchError("failed validating wego installation: Weave GitOps cannot talk to the cluster"))
		})
		Context("return unmodified on GetClusterStatus and valid wego config", func() {
			fakeKubeClient.GetClusterStatusReturns(kube.Unmodified)
			fakeKubeClient.GetWegoConfigReturns(&kube.WegoConfig{
				FluxNamespace: testNamespace,
				WegoNamespace: testNamespace,
			}, nil)
			It("should fail getting cluster name", func() {
				fakeKubeClient.GetClusterNameReturns("", someError)

				err := installer.Install(testNamespace, configRepo, true)
				Expect(err).Should(MatchError(fmt.Sprintf("failed getting cluster name: %s", someError)))
			})
			Context("return valid cluster name", func() {
				fakeKubeClient.GetClusterNameReturns(clusterName, nil)
				It("should fail installing flux", func() {
					fakeFluxClient.InstallReturns(nil, someError)

					err := installer.Install(testNamespace, configRepo, true)
					Expect(err).Should(MatchError(fmt.Sprintf("failed installing flux: %s", someError)))
				})
				Context("do not return error on flux install", func() {
					fakeFluxClient.InstallReturnsOnCall(0, nil, nil)
					It("should fail getting bootstrap manifests", func() {
						fakeFluxClient.InstallReturnsOnCall(1, nil, someError)

						err := installer.Install(testNamespace, configRepo, true)
						Expect(err).Should(MatchError(fmt.Sprintf("failed getting bootstrap manifests: failed getting runtime manifests: %s", someError)))
					})
					Context("do not return error on flux install", func() {
						fakeFluxClient.InstallReturnsOnCall(1, nil, nil)
						It("should fail getting default branch", func() {
							fakeGitProvider.GetDefaultBranchReturnsOnCall(0, "", someError)

							err := installer.Install(testNamespace, configRepo, true)
							Expect(err).Should(MatchError(fmt.Sprintf("failed getting default branch: %s", someError)))
						})
						Context("return valid default branch", func() {
							fakeGitProvider.GetDefaultBranchReturnsOnCall(0, "main", nil)
							It("should fail getting config repo git source", func() {
								fakeGitProvider.GetRepoVisibilityReturns(nil, someError)

								err := installer.Install(testNamespace, configRepo, true)
								Expect(err).Should(MatchError(fmt.Sprintf("failed getting git source: failed getting ref secret: %s", someError)))
							})
							Context("return private repo visibility", func() {
								privateVisibility := gitprovider.RepositoryVisibilityPrivate
								fakeGitProvider.GetRepoVisibilityReturns(&privateVisibility, nil)
								It("should fail applying bootstrap manifests", func() {
									fakeKubeClient.ApplyReturns(someError)

									err := installer.Install(testNamespace, configRepo, true)
									Expect(err).Should(MatchError(fmt.Sprintf("error applying manifest .weave-gitops/clusters/test-cluster/system/wego-system.yaml: %s", someError)))
								})
								Context("do not fail when applying manifests", func() {
									fakeKubeClient.ApplyReturns(nil)
									It("should fail getting gitops manifests", func() {
										fakeFluxClient.InstallReturnsOnCall(2, nil, someError)

										err := installer.Install(testNamespace, configRepo, true)
										Expect(err).Should(MatchError(fmt.Sprintf("failed generating gitops manifests: failed getting runtime manifests: %s", someError)))
									})
									Context("do not fail when calling install for second time", func() {
										fakeFluxClient.InstallReturnsOnCall(2, nil, nil)
										It("should fail writing directly to branch", func() {
											fakeGitClient.CloneReturns(false, someError)

											err := installer.Install(testNamespace, configRepo, true)
											Expect(err).Should(MatchError(fmt.Sprintf("failed writting to default branch failed to clone repo: failed cloning user repo: ssh://git@github.com/test-user/test-repo.git: %s", someError)))
										})
										It("should fail creating a pull requests", func() {
											fakeGitProvider.CreatePullRequestReturns(nil, someError)

											err := installer.Install(testNamespace, configRepo, false)
											Expect(err).Should(MatchError(fmt.Sprintf("failed creating pull request: %s", someError)))
										})
									})
								})
							})
						})
					})
				})
			})
		})
	})
```